2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+> users can stay on PSGet v2 and that will continue to work.
+
+## Motivation
+
+    As a PowerShell user,
+    I can discover how to install missing cmdlets,
+    so that I can be more productive.
+
+    As a PowerShellGet contributor,
+    I can more easily contribute code,
+    so that PowerShellGet evolves more quickly.
+
+## User Experience
+
+```powershell
+PS> Get-Satisfaction
+Get-Satisfaction : The term 'Get-Satisfaction' is not recognized as the name of a cmdlet, function, script file, or operable program.
+Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
+At line:1 char:1
++ Get-Satisfaction
++ ~~~
++ CategoryInfo          : ObjectNotFound: (Get-Satisfaction:String) [], CommandNotFoundException
++ FullyQualifiedErrorId : CommandNotFoundException
+
+Suggestion: You can obtain `Get-Satisfaction` by running `Install-PSResource HappyDays`.
+
+PS> Install-PSResource HappyDays
+```
+
+## Specification
+
+### Rewrite of PowerShellGet
+
+PowerShellGet is currently written as PowerShell script with a dependency on PackageManagement.
+Proposal is to write PSResource in C# to reduce complexity and make easier to maintain.
+In addition, remove dependency on PackageManagement completely as well as dependency on
+nuget.exe.
+This module will only use REST APIs to get and publish nupkgs.
+This module would be shipped in PowerShell 7.
+
+### Side-by-side with PowerShellGet
+
+To ease transition due to the enormity of the breaking change, this module will
+have a different names for the cmdlets to allow for side by side use
+with the existing PowerShellGet module.
+Since the current PowerShellGet 2.x version is a script module and this new one
+is a C# based module, they can coexist side-by-side.
+
+### Cmdlet compatibility
+
+Function wrappers will be provided to provide compatibility with the two most often
+used cmdlets: `Install-Module` and `Find-Module`.
+These cmdlets will output a warning to use the new cmdlets.
+
+### Local cache
+
+Rather than connecting to PSGallery and other repositories for finding modules,
+the user must explicitly request the current metadata of
+modules from the repositories and that is cached locally.
+`Find-PSResource` (see below) works against this local cache.
+This will also enable changes in PowerShell to use `Find-PSResource -Type Command` to look
+in this cache when it can't find a command and suggest to the user the module to install to
+get that command.
+This will be a local json file containing only the latest version of each module.
+The cache will be stored in the user path.
+There is no system cache that is shared.
+
+Example cache entry:
+
+```json
+{
+  "name": "This is my module",
+  "version": "1.0.0.0",
+  "type": "module",
+  "tags": [
+    "Linux",
+    "PSEdition_Core",
+    "PSEdition_Desktop",
+    "AzureAutomation"
+  ]
+}
+```
+
+An example cache with 5000 resources is ~700KB in compressed json form.
```

Comment:
See https://github.com/PowerShell/PowerShellGet/issues/287

Reply:
How would this work with lower version preview/beta/pre-release modules. Would they be stored separately with being pre-release?

Reply:
Anytime it doesn't find it in the cache, it'll have to look online.  This will cover the older version cases which should be the significant minority as most of the time people are looking for latest version.  It may make sense to cache both latest stable and latest preview releases.

Reply:
If 1.3 was the latest version when the cache was downloaded, then 1.4 is published, then I run install-module, what will I get?

Reply:
@edyoung: RFC says install-module would update the cache (first?) right? 😉

Reply:
> It may make sense to cache both latest stable and latest preview releases.

Yes, that's what I was thinking as e.g. the `AzSqlElasticJob` cmdlets in 1.1.1-preview haven't made it into 1.4.0 GA release of the same module yet.

Reply:
I would be in favor of leaving the existing semantics alone for Find-Module/Find-PSresource - always make at least an if-modified-since  request to repo, so you always get latest without having to put update-psresourcecache or -force in your script. The suggestion on command not found could be based on a cache and potentially produce old results because it's a hint/heuristic.

Reply:
This assumes that there could be two request: first one to get manifest timestamp, compare with cache timestamp, then second one to get new manifest if needed.
Or it can be one request with cache timestamp to get delta of the manifest.